spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -474,15 +475,31 @@ public static boolean isLocalityEnabled(FileIO io, String location, CaseInsensit
     return false;
   }
 
-  public static boolean isVectorizationEnabled(Map<String, String> properties, CaseInsensitiveStringMap readOptions) {
+  public static boolean isVectorizationEnabled(FileFormat fileFormat,
+                                               Map<String, String> properties,
+                                               CaseInsensitiveStringMap readOptions) {
     String batchReadsSessionConf = SparkSession.active().conf()
         .get("spark.sql.iceberg.vectorization.enabled", null);
     if (batchReadsSessionConf != null) {
       return Boolean.valueOf(batchReadsSessionConf);
     }
-    return readOptions.getBoolean(SparkReadOptions.VECTORIZATION_ENABLED,
```

Comment:
Can we also make the same change in Spark 2.x DataSourceReader?

Reply:
Good catch, I totally forgot.

Reply:
Actually, I think we don't handle the session config and options correctly. To me, the most specific value is in read options, then in session config, then in table props. If I get it correctly, it is not how it works now. I think we treat the session config as the most important part that takes precedence over anything else.

@shardulm94 @RussellSpitzer @rdblue, do you agree we should change that?

Reply:
I see tradeoffs either way. I agree that the most specific value is ideally the read options explicitly passed to the table read. But a session conf taking higher precedence is also convenient in production to turn off vectorization for an application by a pure config change, no need for code changes.

Another option we have is to use a boolean `AND` between the session conf and read option. This is used in https://github.com/apache/iceberg/blob/91ac42174e4c535ece4e36db2cb587a23babced9/spark2/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java#L182
It can be a little confusing here if the default of session conf (true) is different than the default of read option (false), but is worth considering.

Reply:
Oh, I am a bit surprised how we handle nullability.

My main concern is the session config is usually used for default values that make sense for most tables. However, not having a way to override the session config on a job level does not sound right to me. For example, if a user works with 5 tables in a notebook, I get that he/she won't want to repeat the same configs for all tables. So the session conf is probably a good place for common configs. However, what if there is a config that must be changed for one of those 5 tables? Requiring a separate notebook does not seem reasonable.

Reply:
I agreed with
Read options > Session Options > Table Options

Basically in order of ease of setting

Reply:
I just commented on this above. I think the original order was because read options are compiled into jobs so they are more targeted to a table/scan, but should probably not override runtime config.

Reply:
I guess I always thought about it from the Shell/Notebook perspective, my table options are inside of a property file somewhere and require an alter table command that everyone can see but my read and session options are more or less set at runtime.